src/cluster_legacy.c

```diff
@@ -2887,8 +2890,7 @@ int clusterProcessPacket(clusterLink *link) {
         senderConfigEpoch = ntohu64(hdr->configEpoch);
         if (senderCurrentEpoch > server.cluster->currentEpoch) server.cluster->currentEpoch = senderCurrentEpoch;
         /* Update the sender configEpoch if it is a primary publishing a newer one. */
-        if (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, sizeof(hdr->replicaof)) &&
-            senderConfigEpoch > sender->configEpoch) {
+        if (sender_claims_primary && senderConfigEpoch > sender->configEpoch) {
```

Comment:
Should we move this entire condition to a variable and name it as `sender_is_primary`?

```
int sender_is_primary = (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN) && senderConfigEpoch > sender->configEpoch);
```

Reply:
`sender_is_new_primary` sounds like a better name to me. Let me play with it a bit next.

Reply:
It can be a primary even if `senderConfigEpoch == sender->configEpoch`, can't it?

Only if it was a replica before and a primary now, we can require that the config epoch is bumped.

Reply:
>It can be a primary even if senderConfigEpoch == sender->configEpoch, can't it?

Not sure I follow. The "primary" part is inferred from `memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN)`. And then the epoch check is separate. The idea is to make sure that we are not looking at a previous election result.

>Only if it was a replica before and a primary now, we can require that the config epoch is bumped.

Yes there needs to be a epoch bump for a "promotion". Epoch bump is part of the election protocol than an add-on requirement. 

>Should we move this entire condition to a variable and name it as sender_is_primary?

The new name `sender_claims_to_be_primary` is a lot clearer now and also very long. I haven't been able to come up with a better name that would allow me to encode the "more recent" property, which would probably read like `sender_cliams_to_be_a_more_recent_primary_from_my_perspective` :D. I think I will keep `sender_claims_to_be_primary` for now.

Reply:
> > It can be a primary even if senderConfigEpoch == sender->configEpoch, can't it?
> 
> Not sure I follow.

This was in response to @hpatro's suggestion above, which includes `senderConfigEpoch > sender->configEpoch`:

    int sender_is_primary = (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN) && senderConfigEpoch > sender->configEpoch);